Remotion/Data/DomainObjects/Mapping/Builder/ClassDefinitionBuilderNode.cs

```diff
@@ -0,0 +1,111 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Remotion.Utilities;
+
+namespace Remotion.Data.DomainObjects.Mapping.Builder
+{
+  /// <summary>
+  /// Represents <see cref="Mapping.ClassDefinition"/> node in a tree constructed by <see cref="TypeDefinitionFactory"/>.
+  /// Supports child navigation and correctly setting up child relationships in a resulting <see cref="Mapping.ClassDefinition"/>.
+  /// </summary>
+  public class ClassDefinitionBuilderNode : TypeDefinitionBuilderNode
+  {
+    public ClassDefinitionBuilderNode? BaseClass { get; }
+
+    public ClassDefinition? ClassDefinition { get; private set; }
+
+    public bool IsReadOnly { get; private set; }
+
+    private readonly List<ClassDefinitionBuilderNode> _derivedClasses = new();
+
+    public ClassDefinitionBuilderNode (Type type, ClassDefinitionBuilderNode? baseClass)
+        : base(type)
+    {
+      if (!type.IsClass)
+        throw new ArgumentException("The specified type must be a class.", "type");
+      if (baseClass?.IsReadOnly == true)
+        throw new ArgumentException("The specified base class must not be read-only.", "baseClass");
+
+      BaseClass = baseClass;
+      BaseClass?.AddDerivedClass(this);
+    }
+
+    public IReadOnlyList<ClassDefinitionBuilderNode> DerivedClasses => _derivedClasses;
+
+    /// <inheritdoc />
+    public override bool IsLeafNode => DerivedClasses.Count == 0;
+
+    /// <inheritdoc />
+    protected override void BuildTypeDefinition (IMappingObjectFactory mappingObjectFactory)
+    {
+      ArgumentUtility.CheckNotNull("mappingObjectFactory", mappingObjectFactory);
+
+      if (IsReadOnly)
+        throw new InvalidOperationException("Cannot build type definition as the builder node is read-only.");
+
+      if (ClassDefinition != null)
+        return;
+
+      ClassDefinition? baseClass = null;
+      if (BaseClass != null && !ReflectionUtility.IsInheritanceRoot(Type))
+      {
+        BaseClass.BuildTypeDefinition(mappingObjectFactory);
+        baseClass = BaseClass.ClassDefinition;
+      }
+
+      ClassDefinition = mappingObjectFactory.CreateClassDefinition(Type, baseClass);
+    }
+
+    /// <inheritdoc />
+    public override void EndBuildTypeDefinition ()
+    {
+      if (IsReadOnly)
+        return;
+
+      IsReadOnly = true;
+
+      BaseClass?.EndBuildTypeDefinition();
+      foreach (var derivedClass in DerivedClasses)
+        derivedClass.EndBuildTypeDefinition();
+
+      if (ClassDefinition != null)
+        ClassDefinition.SetDerivedClasses(DerivedClasses.Select(e => e.ClassDefinition).Where(e => e is { BaseClass: { } })!);
```

Comment:
This is the problem with LINQ queries and nullability. A `Where` does not change the nullability of the enumerable. So nullable in, nullable out. This is sad because we have the `Where` null check but that won't work. A similar problem happens with the proposed split in the first comment.

We could introduce a `.WhereNotNull` and maybe `.WhereNotNullBy` extension method to support nullability in such scenarios. Otherwise we need to use the `!`.

Reply:
okay, you can overrode nullability for the generic parameter by overriding the nullability of the entire result. This isn't the first time we have null-conversions in linq queries. The other times, it was done by appending `.Select(e => e!)` (e.g. ReflectionBasesPropertyResolver::66). I would like to keep this pattern to make the null override with the least impact. I don't mind if we one day introduce a `WhereNotNull` filter. Right now, there's about 10 instances of this in re-motion.